libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/trivially_copyable.compile.pass.cpp

```diff
@@ -0,0 +1,32 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// REQUIRES: c++20 || clang
+
+#include <type_traits>
+#include <utility>
+
+struct trivially_copyable {
+  int arr[4];
+};
+
+static_assert(std::is_trivially_copyable<std::pair<int, int>>::value, "");
+static_assert(std::is_trivially_copyable<std::pair<int, char>>::value, "");
+static_assert(std::is_trivially_copyable<std::pair<char, int>>::value, "");
+static_assert(std::is_trivially_copyable<std::pair<std::pair<char, char>, int>>::value, "");
+static_assert(std::is_trivially_copyable<std::pair<trivially_copyable, int>>::value, "");
+
+static_assert(!std::is_trivially_copyable<std::pair<int&, int>>::value, "");
```

Comment:
It's possible to make `std::pair<int&, int>` trivially copyable and MSVC STL is already doing so ([demo](https://godbolt.org/z/fbjoEcb8G)).

MSVC STL's `pair` has a [deleted copy assignment operator taking `const volatile pair&`](https://github.com/microsoft/STL/blob/0515a05b394596de92d08cb0f352614479a2a883/stl/inc/utility#L314), and its usable assignment operators (e.g. [the one performing copy assignment](https://github.com/microsoft/STL/blob/0515a05b394596de92d08cb0f352614479a2a883/stl/inc/utility#L316-L320)) are templates. The usable assigment operators are non-trivial, but not counted for copy/move assignment operators, so the `pair` specialization can still be trivially copyable.

Perhaps we can adopt such technique when either `first_type` or `second_type` is a reference type.

Reply:
I'm not convinced this is an especially useful property. `pair`s with references in them is already weird, and this feels a bit like lying to the user. When do you care that your pair with a reference is trivially copyable?